akamai/cache-client.go

```diff
@@ -0,0 +1,255 @@
+)
+
+const (
+	purgePath       = "/ccu/v2/queues/default"
+	timestampFormat = "20060102T15:04:05-0700"
+)
+
+type purgeRequest struct {
+	Objects []string `json:"objects"`
+}
+
+type purgeResponse struct {
+	HTTPStatus       int    `json:"httpStatus"`
+	Detail           string `json:"detail"`
+	EstimatedSeconds int    `json:"estimatedSeconds"`
+	PurgeID          string `json:"purgeId"`
+}
+
+// CachePurgeClient talks to the Akamai CCU REST API. It is safe to make concurrent
+// purge requests.
+type CachePurgeClient struct {
+	client       *http.Client
+	apiEndpoint  string
+	apiHost      string
+	apiScheme    string
+	clientToken  string
+	clientSecret string
+	accessToken  string
+	issuer       *x509.Certificate
+	retries      int
+	retryBackoff time.Duration
+	log          *blog.AuditLogger
+	stats        statsd.Statter
+	clk          clock.Clock
+}
+
+type errFatal string
+
+func (e errFatal) Error() string { return string(e) }
+
+var (
+	// ErrAllRetriesFailed lets the caller of Purge to know if all the purge submission
+	// attempts failed
+	ErrAllRetriesFailed = errors.New("All attempts to submit purge request failed")
+)
+
+// NewCachePurgeClient constructs a new CachePurgeClient
+func NewCachePurgeClient(
+	endpoint,
+	clientToken,
+	clientSecret,
+	accessToken string,
+	retries int,
+	retryBackoff time.Duration,
+	issuer *x509.Certificate,
+	log *blog.AuditLogger,
+) (*CachePurgeClient, error) {
+	apiURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	return &CachePurgeClient{
+		client:       new(http.Client),
+		apiEndpoint:  endpoint,
+		apiHost:      apiURL.Host,
+		apiScheme:    strings.ToLower(apiURL.Scheme),
+		clientToken:  clientToken,
+		clientSecret: clientSecret,
+		accessToken:  accessToken,
+		retries:      retries,
+		retryBackoff: retryBackoff,
+		log:          log,
+		clk:          clock.Default(),
+		issuer:       issuer,
+	}, nil
+}
+
+func (cpc *CachePurgeClient) constructAuthHeader(request *http.Request, body []byte, apiPath string, nonce string) (string, error) {
+	timestamp := cpc.clk.Now().Format(timestampFormat)
+	header := fmt.Sprintf(
+		"EG1-HMAC-SHA256 client_token=%s;access_token=%s;timestamp=%s;nonce=%s;",
+		cpc.clientToken,
+		cpc.accessToken,
+		timestamp,
+		nonce,
+	)
+	bodyHash := sha256.Sum256(body)
+	tbs := fmt.Sprintf(
+		"%s\t%s\t%s\t%s\t%s\t%s\t%s",
+		"POST",
+		cpc.apiScheme,
+		cpc.apiHost,
+		apiPath,
+		"", // We don't need to send any signed headers so this can be blank
+		base64.StdEncoding.EncodeToString(bodyHash[:]),
+		header,
+	)
+
+	h := hmac.New(sha256.New, []byte(cpc.clientSecret))
+	h.Write([]byte(timestamp))
+	key := base64.StdEncoding.EncodeToString(h.Sum(nil))
+
+	h = hmac.New(sha256.New, []byte(key))
+	h.Write([]byte(tbs))
+	sig := base64.StdEncoding.EncodeToString(h.Sum(nil))
+
+	return fmt.Sprintf("%ssignature=%s", header, sig), nil
+}
+
+func (cpc *CachePurgeClient) generateURLs(der []byte) ([]string, error) {
```

Comment:
This code just needs to know how to talk to akamai. We can do the DER work above it.